DataStructures/Lists/CircularBuffer.cs

```diff
@@ -0,0 +1,197 @@
+        /// </summary>
+        /// <param name="value">value to be added to the buffer</param>
+        public void Add(T value) 
+        {
+            if (CanOverride) 
+            {
+                innerInsert(value);
+            }
+            else 
+            {
+                if (IsFilledUp) 
+                {
+                    throw new CircularBufferFullException($"Circular Buffer is filled up. {value} can not be inserted");
+                }
+                innerInsert(value);
+            }
+        }
+
+        // Inserts data into the buffer without checking if it is full
+        private void innerInsert(T value) 
+        {
+            _circularBuffer[_end] = value;
+            _end = (_end + 1) % _circularBuffer.Length;
+            if (_end == _start) 
+            {
+                _start = (_start + 1) % _circularBuffer.Length;
+            }
+        }
+
+        /// <summary>
+        ///     Reads and removes the value in front of the buffer, and places the next value in front.
+        /// </summary>
+        public T Pop() 
+        {
+            var result = _circularBuffer[_start];
+            _circularBuffer[_start] = _circularBuffer[_end];
+            _start = (_start + 1) % _circularBuffer.Length;
+            return result;
+        }
+
+        #region IEnumerable Implementation
+        public IEnumerator<T> GetEnumerator() 
+        {
+            for (int i = 0; i < Count; i++) {
+                yield return _circularBuffer[i];
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() 
+        {
+            return GetEnumerator();
+        }
+        #endregion
+
+        #region ICollection Implementation
+        /// <summary>
+        /// Returns the number of elements.
+        /// </summary>
+        public int Count 
+        {
+            get { return _length; }
+        }
+        /// <summary>
+        /// Checks whether this collection is readonly
+        /// </summary>
+        public bool IsReadOnly 
+        {
+            get { return false; }
+        }
+        /// <summary>
+        /// Clears this instance
+        /// </summary>
+        public void Clear() 
+        {
+            for (int i = 0; i < _circularBuffer.Length; i++) 
+            {
+                _circularBuffer[i] = _default;
+            }
+        }
+        /// <summary>
+        /// Checks whether the buffer contains an item
+        /// </summary>
+        public bool Contains(T item) 
+        {
+            return _circularBuffer.Contains(item);
+        }
+        /// <summary>
+        /// Copies this buffer to an array
+        /// </summary>
+        public void CopyTo(T[] array, int arrayIndex) 
+        {
+            if (array == null) 
+            {
+                throw new ArgumentNullException("array can not be null");
+            }
+
+            if (array.Length == 0 || arrayIndex >= array.Length || arrayIndex < 0) {
```

Comment:
Inconsistent formatting, sometimes you place the `{` on the same line and sometimes you place it on a new line. Please place all `{`, `}` characters on new lines.

Reply:
Sorry for that. I came from the Java side of things. It's all consistent now

Reply:
Sorry about that. I came from the Java side of things. It's all consistent now